design/one-pager-managed-resource-api-design.md

```diff
@@ -0,0 +1,294 @@
+  reclaimPolicy: Delete
+  writeConnectionSecretToRef: {}
+  forProvider:
+    name: monus-legacy-network-res
+    routingConfig:
+      routingMode: REGIONAL
+status:
+  conditions:
+  - lastTransitionTime: "2019-08-29T05:55:16Z"
+    reason: Successfully reconciled managed resource
+    status: "True"
+    type: Synced
+  atProvider:
+    IPv4Range: 10.240.0.0/16
+    creationTimestamp: "2019-08-28T10:45:27.445-07:00"
+    gatewayIPv4: 10.240.0.1
+    id: 75666386684937048
+    kind: compute#network
+    name: monus-legacy-network-res
+    routingConfig:
+      routingMode: REGIONAL
+    selfLink: https://www.googleapis.com/compute/v1/projects/crossplane-playground/global/networks/monus-legacy-network-res
+```
+
+### Pointer Types and Markers
+
+Related to https://github.com/crossplaneio/crossplane/issues/741
+
+For any field, the developer can choose whether to use pointer type or value type, i.e. `*bool` vs `bool`. The main
+difference here is their zero-value, which gets used when the field is left empty. We follow Kubernetes conventions
+for this decision: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required
+
+Short summary:
+* Optional fields are pointer type -> `*AwesomeType` and marked as optional via a comment marker `//+optional` and
+`omitempty` struct tag.
+* Required fields are exact opposite, i.e. not pointer type, not marked as `//+optional` and do not have `omitempty`
+struct tag.
+
+Here is the flow to decide how to make the decision for `Spec` fields:
+1. Start with assuming all fields are optional.
+2. Convert the fields to required if any of CRUD calls you make requires them, i.e. it should be possible to call all
+CRUD operations with all those required fields filled.
+
+Note that some required fields by all CRUD calls might be late-initalized. For example, AWS VPC id can only be fetched
+from the provider and CRUD operations except `Create` requires it. In those cases, mark the field as optional since at
+any step we may not have it.
+
+> By pointer types, developer tries to show that the value of the field may not exist and zero-value defaulting does not
+make sense for that type.
+
+### Immutable Properties
+
+Related to https://github.com/crossplaneio/crossplane/issues/727
+
+Some of the fields that include in `Spec` can only be configured in creation call of the resource, later you cannot
+update them. However, Kubernetes Custom Resources validation mechanisms do not yet support that behavior, see https://github.com/kubernetes/enhancements/pull/1099
+for details.
+
+There are some solutions like admission webhooks to enforce immutability of some of the fields, however, current behavior
+is that Crossplane shows the user what the error is received when the call that includes a change in the immutable
+property is made and the most up-to-date status under `Status`. In this scenario, we have to make sure that all immutable
```

Comment:
> Crossplane shows the user what the error is received when the call that includes a change in the immutable
property is made

I wish this were true, but I'm pretty sure it's worse in many cases. Some cloud provider APIs have distinct endpoints (and thus Go structs) for update operations, so it's possible that when the user edits a field the provider considers immutable we'll just never ask that it be updated. ReplicationGroup is an example of this. If someone edited a field that wasn't covered in `NewModifyReplicationGroupInput` we'd just happily accept the change to our custom resource and do nothing:

https://github.com/crossplaneio/stack-aws/blob/b64415a/pkg/controller/cache/managed.go#L217
https://github.com/crossplaneio/stack-aws/blob/b64415a/pkg/clients/elasticache/elasticache.go#L142

I think a good first step toward preventing this would be to require managed resource authors include the `//+immutable` comment marker that is proposed by https://github.com/kubernetes/enhancements/pull/1099. While the KEP is up for debate we can at least teach our API documentation generator to document that such fields are immutable, and if we're lucky we'll get CRD based immutability for free in some future version of Kubernetes. 😬

Reply:
Yeah, I didn't take that case into account. I guess the best we can do for now to have `//+immutable` and hope one day we either have admission webhooks generated using this marker or native k8s support to actually incorporate it.

> If someone edited a field that wasn't covered in NewModifyReplicationGroupInput we'd just happily accept the change to our custom resource and do nothing

This upset me a little.